xlators/mgmt/glusterd/src/glusterd-pmap.c

```diff
@@ -152,275 +188,187 @@ pmap_registry_search(xlator_t *this, const char *brickname,
     return 0;
 }
 
-static int
-pmap_registry_search_by_xprt(xlator_t *this, void *xprt,
-                             gf_pmap_port_type_t type)
+int
+pmap_registry_search_by_xprt(xlator_t *this, void *xprt)
 {
     struct pmap_registry *pmap = NULL;
-    int p = 0;
+    struct pmap_ports *tmp_port = NULL;
     int port = 0;
 
     pmap = pmap_registry_get(this);
-
-    for (p = pmap->last_alloc; p >= pmap->base_port; p--) {
-        if (!pmap->ports[p].xprt)
-            continue;
-        if (pmap->ports[p].xprt == xprt) {
-            if (pmap->ports[p].type == type || type == GF_PMAP_PORT_ANY) {
-                port = p;
-                break;
-            }
+    cds_list_for_each_entry(tmp_port, &pmap->ports, port_list)
+    {
+        if (tmp_port->xprt == xprt) {
+            port = tmp_port->port;
+            break;
         }
     }
 
     return port;
 }
 
-static char *
-pmap_registry_search_by_port(xlator_t *this, int port)
-{
-    struct pmap_registry *pmap = NULL;
-    char *brickname = NULL;
-    int max_port = 0;
-
-    max_port = ((glusterd_conf_t *)(this->private))->max_port;
-    if (port > max_port)
-        goto out;
-
-    pmap = pmap_registry_get(this);
-
-    if (pmap->ports[port].type == GF_PMAP_PORT_BRICKSERVER)
-        brickname = pmap->ports[port].brickname;
-
-out:
-    return brickname;
-}
-
 int
-pmap_registry_alloc(xlator_t *this)
+port_brick_bind(xlator_t *this, int port, char *brickname, void *xprt)
 {
     struct pmap_registry *pmap = NULL;
-    int p = 0;
-    int port = 0;
+    struct pmap_ports *tmp_port = NULL;
+    char *tmp_brick;
+    int ret = -1;
 
-    pmap = pmap_registry_get(this);
+    GF_ASSERT(this);
 
-    for (p = pmap->base_port; p <= pmap->max_port; p++) {
-        /* GF_PMAP_PORT_FOREIGN may be freed up ? */
-        if ((pmap->ports[p].type == GF_PMAP_PORT_FREE) ||
-            (pmap->ports[p].type == GF_PMAP_PORT_FOREIGN)) {
-            if (pmap_port_isfree(p)) {
-                pmap->ports[p].type = GF_PMAP_PORT_LEASED;
-                port = p;
-                break;
-            }
+    pmap = pmap_registry_get(this);
+    cds_list_for_each_entry(tmp_port, &pmap->ports, port_list)
+    {
+        if (tmp_port->port == port) {
+            ret = 0;
+            break;
         }
     }
 
-    if (port > pmap->last_alloc)
-        pmap->last_alloc = port;
+    if (ret) {
+        ret = pmap_add_port_to_list(this, port, brickname, xprt);
+        if (ret)
+            gf_msg(this->name, GF_LOG_ERROR, errno, GD_MSG_BRICK_ADD_FAIL,
+                   "Failed to add brick to the ports list");
+    } else {
+        tmp_brick = tmp_port->brickname;
+        ret = gf_asprintf(&tmp_port->brickname, "%s %s", tmp_brick, brickname);
```

Comment:
I think we need to old brickname from tmp_port before save a new brickname.

Reply:
I think new logic is not similar to previous one, we need to change this,Let's suppose we have an entry for a brickname in port list and we do call this function again it simply replace an entryname without ensuring entry is same or not.This behavior can create a big issue, we need to persist the behavior similar to older one like this
old_bn = pmap->ports[port].brickname;	
    if (old_bn) {	
        bn_len = strlen(brickname);	
        entry = strstr(old_bn, brickname);	
        while (entry) {	
            found = 1;	
            if ((entry != old_bn) && (entry[-1] != ' ')) {	
                found = 0;	
            }	
            if ((entry[bn_len] != ' ') && (entry[bn_len] != '\0')) {	
                found = 0;	
            }	
            if (found) {	
                return 0;	
            }	
            entry = strstr(entry + bn_len, brickname);	
        }	
        asprintf(&new_bn, "%s %s", old_bn, brickname);	
    } else {	
        new_bn = strdup(brickname);	
    }

Reply:
@nik-redhat We need to run regress testing to validate the patch, the portmap specific code path is sensitive and we use this to take decision about attach/detach brick so a small condition break in this code path it can create a big issue.

Reply:
Sure.....I will update the mentioned changes, do testing of the same, and update here.